google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -174,18 +174,18 @@ def _process_requests(
         # Handle special errors returned for ack/modack RPCs via the ErrorInfo
         # sidecar metadata when exactly-once delivery is enabled.
         if errors_dict and ack_id in errors_dict:
-            exactly_once_error = errors_dict[ack_id]
+            exactly_once_error = str(errors_dict[ack_id])
             if exactly_once_error.startswith("TRANSIENT_"):
                 requests_to_retry.append(ack_reqs_dict[ack_id])
             else:
                 if exactly_once_error == "PERMANENT_FAILURE_INVALID_ACK_ID":
                     exc = AcknowledgeError(AcknowledgeStatus.INVALID_ACK_ID, info=None)
                 else:
                     exc = AcknowledgeError(AcknowledgeStatus.OTHER, exactly_once_error)
-
                 future = ack_reqs_dict[ack_id].future
-                future.set_exception(exc)
-                requests_completed.append(ack_reqs_dict[ack_id])
+                if future is not None:
+                    future.set_exception(exc)
+                    requests_completed.append(ack_reqs_dict[ack_id])
```

Comment:
`requests_completed.append(ack_reqs_dict[ack_id])` should be un-indented. We want to complete these requests. Completed requests are removed from lease management by the dispatcher (it calls drop on these.) If we didn't complete them, we'd lose track of these requests. (The dispatcher expects all passed in ack/modack requests to be in either requests_completed or requests_to_retry; as the code stands, the request wouldn't be returned in either list.) The request shouldn't be retried b/c there was a permanent error returned for this ack id.